File: codejam/alien.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <set>
using namespace std;

int main() {

  ifstream fin("input.txt");

  int length, size, cases;
  fin >> length;
  fin >> size;
  fin >> cases;

  string words[size];
  string input;

  //cout << size << endl;
  for(int i = 0; i < size; i++) {
    fin >> words[i];
  }

  // for(int i = 0; i < size; i++) {
  //   cout << i << " : ";
  //   cout << words[i] << endl;
  // }

  while(cases--) {
    fin >> input;
    cout << input << endl;
    
    int pos = 0;
    int ans = 0;
    set<int> location, temp;
    for(int i = 0; i < size; i++) location.insert(i);

    for(int i = 0; i < input.length(); ++i) {
      int iter = (int) location.size();
      if(input[i] == '(') {
  	++i;
  	while(input[i] != ')') {
  	  set<int>::iterator it = location.begin();
  	  for(int j = 0; j < iter; ++it) {
  	    if(words[*it][pos] == input[i])
  	      temp.insert(*it);
  	  }
  	  ++i;
  	}
      }
      else {
  	set<int>::iterator it = location.begin();
  	for(int j = 0; j < iter; ++it) {
  	  if(words[*it][pos] == input[i])
  	    temp.insert(*it);
  	}
      }
      location.clear();
      location = temp;
      temp.clear();
      pos++;
      ans = (int) location.size();
    }
    cout << ans;
  }
    
  return 0;
}

File: codejam/alien.py
#!/usr/bin/env python

f = open('input.txt', 'r+')

tmp = f.readline()
tmp = tmp.split(' ')
tmp = [int(x) for x in tmp]
length, size, cases = tmp[0], tmp[1], tmp[2]
words = []

for n in range(0, size):
    words.append(f.readline()[:-1])

word = words[:]
for n in range(0, cases):
    case = f.readline()[:-1]
    pos = 0
    possible = word[:]
    x = 0
    first = True
    while x < len(case):
        if first:
            possible = word[:]
            first = False
        else:
            possible = list(words)[:]
        words = set([])
        if case[x] == '(':
            x += 1
            while case[x] != ')':
                letter = case[x]
                for w in possible:
                    if w[pos] == letter:
                        words.add(w)
                x += 1
            x += 1
        else:
            for w in possible:
                if w[pos] == case[x]:
                    words.add(w)

            x += 1
        pos += 1
        
    print "Case #" + str(n+1) + ":", len(words)
